Universal/Sniffs/CodeAnalysis/MixedBooleanOperatorSniff.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * PHPCSExtra, a collection of sniffs and standards for use with PHP_CodeSniffer.
+ *
+ * @package   PHPCSExtra
+ * @copyright 2021 WoltLab GmbH, 2023 PHPCSExtra Contributors
+ * @license   https://opensource.org/licenses/LGPL-3.0 LGPL3
+ * @link      https://github.com/PHPCSStandards/PHPCSExtra
+ */
+
+namespace PHPCSExtra\Universal\Sniffs\CodeAnalysis;
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Util\Tokens;
+use PHPCSUtils\BackCompat\BCFile;
+
+/**
+ * Forbid mixing different binary boolean operators within a single expression without making precedence
+ * clear using parentheses.
+ *
+ * @link https://github.com/squizlabs/PHP_CodeSniffer/pull/3205
+ * @link https://github.com/php-fig/per-coding-style/issues/22
+ *
+ * @since 1.2.0
+ */
+final class MixedBooleanOperatorSniff implements Sniff
+{
+    private $previousTokens = [];
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @since 1.2.0
+     *
+     * @return array<int|string>
+     */
+    public function register()
+    {
+        $this->previousTokens = \array_merge(
+            Tokens::$booleanOperators,
+            [\T_INLINE_THEN, \T_INLINE_ELSE]
+        );
```

Comment:
That's fair, but that doesn't answer my question about the use of `array_merge()`.

Reply:
Actually, no, there is one more alternative you seem to overlook (which complies with your "don't mutate" principle):
```php
        $this->previousTokens = Tokens::$booleanOperators + [\T_INLINE_THEN => \T_INLINE_THEN, \T_INLINE_ELSE => \T_INLINE_ELSE];
```

Reply:
Okay, so maybe I should have made the question more specific: why `array_merge()` instead of array union ?

Reply:
The semantics of applying the `+` operator to arrays (“Array Union”) are so terrible (especially with regard to arrays that satisfy `array_is_list()`) that I don't believe there is ever a good reason to use it, because it makes the code less clear.

Reply:
FYI: I will add a new token collection to PHPCSUtils in the 1.1.0 version for the ternary operator tokens.

Reply:
> Okay, so I fundamentally disagree with you on this. Array merge and array union both have a valid use case if used correctly.

I'm not surprised that exposure to different code-bases will cause folks to have different opinions with regard to how stuff should or should not look like. I'm generally dealing with either lists (i.e. numbered indices from 0 to n without holes) or maps (i.e. string keys) and in those cases `array_merge()` does exactly the right thing.

> it is still an anti-patterns which I don't want to see in sniffs in this code base.

Yeah, I'm most certainly not here to discuss your coding style (if you wouldn't have explicitly asked your question) and have now shamelessly stolen the desired style from #277.